lib/rops/index.js

```diff
@@ -0,0 +1,113 @@
+const through = require('through2');
+const csv = require('csv-stringify');
+const AWS = require('aws-sdk');
+const archiver = require('archiver');
+
+module.exports = ({ models, s3 }) => {
+
+  const { Establishment, Project, Procedure } = models;
+
+  const S3 = new AWS.S3({
+    apiVersion: '2006-03-01',
+    region: s3.region,
+    accessKeyId: s3.accessKey,
+    secretAccessKey: s3.secret
+  });
+
+  return ({ id, key }) => {
+
+    return new Promise((resolve, reject) => {
+
+      const establishments = csv({ header: true });
+      const projects = csv({ header: true });
+      const procedures = csv({ header: true });
+
+      const meta = {
+        due: 0,
+        submitted: 0
+      };
+
+      Establishment.query()
+        .select(
+          'id',
+          'name',
+          'licenceNumber',
+          'address',
+          'country'
+        )
+        .toKnexQuery()
+        .stream()
+        .pipe(through.obj((record, enc, callback) => callback(null, record)))
+        .pipe(establishments)
+        .on('error', reject);
+
+      Project.query()
+        .select(
+          'rops.id',
+          'projects.licenceNumber',
+          'projects.establishmentId',
+          'projects.title',
+          'projects.status as projectStatus',
+          'projects.issueDate',
+          'projects.expiryDate',
+          'projects.revocationDate',
+          'licenceHolder.firstName',
+          'licenceHolder.lastName',
+          'licenceHolder.email',
+          'licenceHolder.telephone',
+          'rops.*'
+        )
+        .joinRelated('licenceHolder')
+        .leftJoinRelated('rops')
+        .whereRopsDue(key)
+        .where(q => q.where('rops.year', key).orWhereNull('rops.year'))
```

Comment:
What does this add on top of the `whereRopsDue(year)` ?

Reply:
My understanding of `.whereRopsDue(year)` is that it filters the _projects_ table to ones with a ROP due in that year. This is then adding an extra filter to the `rops` associated with those projects.

Without this, you potentially get multiple rows for each project, with one per existing ROP for any year as long as the project itself was due a ROP in that year.

Reply:
I think there's a `withRops(year)` query builder function on Projects schema, that might do what you're after as well.

Reply:
Although that won't include null year, so maybe not.

Reply:
I'll check it out. As long as it doesn't mind if there is no ROP at all then it looks like the guy.

Reply:
It needs to be as it is above. Using `.withRops` caused multiple rows for a single project when there were multiple ROPs for that project across different years.